### PR TITLE
github: Move windows CI builds to GitHub Actions.

### PR DESCRIPTION
**Pull request by dlech**
By moving to GitHub actions, all MicroPython CI builds are now on GitHub actions. This allows faster parallel builds and saves time by not building when no relevant files changed.

This reveals a few failing tests, so those are temporarily disabled until they can be fixed.

The `misc_sys_settrace_features.py` is disable on visual studio build because it is failing with:

```
--- "D:/a/micropython/micropython/tests/results\\misc_sys_settrace_features.py.exp"	2022-12-17 23:24:57.940564100 +0000
+++ "D:/a/micropython/micropython/tests/results\\misc_sys_settrace_features.py.out"	2022-12-17 23:24:57.940564100 +0000
@@ -40,799 +40,17 @@
  1:   @__main__:factorial => sys_settrace_features.py:91
  2:   @__main__:factorial => sys_settrace_features.py:91
  3:   @__main__:do_tests => sys_settrace_features.py:97
- 4:   @__main__:<module> => sys_settrace_features.py:106
-### trace_handler::main event: line
- 0:   @__main__:factorial => sys_settrace_features.py:88
- 1:   @__main__:factorial => sys_settrace_features.py:91
- 2:   @__main__:factorial => sys_settrace_features.py:91
- 3:   @__main__:do_tests => sys_settrace_features.py:97
- 4:   @__main__:<module> => sys_settrace_features.py:106
[snipped repeating]

------------------- script exited ------------------
-Total traces executed:  155
+Traceback (most recent call last):
+  File "misc/sys_settrace_features.py", line 106, in <module>
+  File "misc/sys_settrace_features.py", line 97, in do_tests
+  File "misc/sys_settrace_features.py", line 91, in factorial
+  File "misc/sys_settrace_features.py", line 91, in factorial
+  File "misc/sys_settrace_features.py", line 69, in trace_tick_handler
+  File "misc/sys_settrace_features.py", line 38, in trace_tick
+  File "misc/sys_settrace_features.py", line 30, in print_stacktrace
+  File "misc/sys_settrace_features.py", line 30, in print_stacktrace
+  File "misc/sys_settrace_features.py", line 30, in print_stacktrace
+  File "misc/sys_settrace_features.py", line 30, in print_stacktrace
+  File "misc/sys_settrace_features.py", line 25, in print_stacktrace
+RuntimeError: maximum recursion depth exceeded
+CRASH
\ No newline at end of file

FAILURE D:/a/micropython/micropython/tests/results\misc_sys_settrace_features.py
```

I think we didn't see this one before because Appveyor builds only did a Debug build for the standard variant but now we are only doing Debug build for the dev variant.

The `import/import_file.py` is disabled on msys builds because it is failing with:

```
 --- /d/a/micropython/micropython/tests/results/import_import_file.py.exp	2022-12-18 00:30:45.784368200 +0000
+++ /d/a/micropython/micropython/tests/results/import_import_file.py.out	2022-12-18 00:30:45.784368200 +0000
@@ -1 +1 @@
-D:\a\micropython\micropython\tests\import\import1b.py
+D:/a/micropython/micropython/tests/import/import1b.py

FAILURE /d/a/micropython/micropython/tests/results/import_import_file.py
```

On the bright side, the failing tests show that the conditional "Print failures" work as expected.

- https://github.com/dlech/micropython/actions/runs/3722233120/jobs/6312902439#step:8:823
- https://github.com/dlech/micropython/actions/runs/3722446041/jobs/6313276996#step:9:11

### CODE REVIEW

**Comment by jimmo** [comment]
Thanks @dlech this sounds really good. How easy is it to debug failing builds on thr GH Actions Windows infrastructure? One of the near features of Appveyor is that you could RDP into the VM for an hour after the build ran, which was quite handy for scenarios that were difficult to repro locally.

File: .github/workflows/ports_windows.yml [hunk 1]
```diff
@@ -17,7 +17,107 @@ concurrency:
...
+  build-vs:
+    strategy:
+      matrix:
+        architecture: [x86, x64]
+        configuration: [Debug, Release]
+        variant: [dev, standard]
+        exclude:
+        - variant: standard
+          configuration: Debug
+    runs-on: windows-2022
```
Comment (stinos) [inline .github/workflows/ports_windows.yml:29]: The 2022 probably means VS2022 right? Appveyor was still running VS2017 intentionally to make sure we're somewhat backwards compatible.

Comment (dlech) [reply .github/workflows/ports_windows.yml:29]: Yes, VS2022 is what is installed on this image by default. We can update this to install an older version if needed or even add multiple versions of VS to the build matrix.

**Comment by dlech** [comment]
> How easy is it to debug failing builds on the GH Actions Windows infrastructure?

The same as Linux I guess. I've always been able to reproduce issues locally so I've never look any farther into it.

File: .github/workflows/ports_windows.yml (hunk 1 above)
Comment (dpgeorge) [reply .github/workflows/ports_windows.yml]: I see that this PR's been updated to test 2017, 2019 and 2022.  That's great!

**Comment by jimmo** [comment]
> > How easy is it to debug failing builds on the GH Actions Windows infrastructure?
> 
> The same as Linux I guess. I've always been able to reproduce issues locally so I've never look any farther into it.

Generally I haven't had too much trouble reproducing Linux issues remotely (and with Docker I've been able to very easily replicate the Actions environment).

But for Windows this just isn't an option. I do (some of the time) have access to a Windows laptop, but certainly not one with all the different VS versions installed etc. So being able to RDP to the Appveyor instance has been invaluable.

**Comment by dlech** [comment]
Seems like it is at least technically possible to do something like this with github actions: https://medium.com/the-scale-factory/troubleshoot-github-actions-via-vpn-3d5d41b01ea4

File: .github/workflows/ports_windows.yml [hunk 2]
```diff
@@ -17,7 +17,130 @@ concurrency:
...
+    steps:
+    - name: Install Visual Studio 2017
+      if: matrix.visualstudio == '2017'
+      run: |
+        choco install visualstudio2017buildtools
+        choco install visualstudio2017-workload-vctools
+        choco install windows-sdk-8.1
+    - uses: microsoft/setup-msbuild@v1
+      with:
+        vs-version: ${{ matrix.vs_version }}
```
Comment (stinos) [inline .github/workflows/ports_windows.yml:51]: Just for my information: are these spaces in `{{}}` required?

File: .github/workflows/ports_windows.yml [hunk 3]
```diff
@@ -17,7 +17,130 @@ concurrency:
...
+    - uses: actions/setup-python@v4
+      if: matrix.runner == 'windows-2019'
+      with:
+        python-version: '3.9'
```
Comment (stinos) [inline .github/workflows/ports_windows.yml:55]: Appveyor used to have 3.8,  any reason for 3.9 ?

**Comment by stinos** [comment]
Overall I like the idea of github actions for everything, but I also think we need at least a POC that remote login is possible. Or even a script on the wiki or so. Especially for mingw builds I've had it more than once that locally things worked ok but then failed on Appveyor because of a slightly different version or so. In which case RDP is usually quicker than trying to reproduce the exact same environment. An alternative would be to make it easier to reproduce the same environment I guess. Maybe a step which prints out versions of everything used explicitly (mingw/gcc/python) ?

File: .github/workflows/ports_windows.yml [hunk 4]
```diff
@@ -17,7 +17,130 @@ concurrency:
...
+          vs_version: '[16, 17)'
+        - visualstudio: '2022'
+          runner: windows-2022
+          vs_version: '[17, 18)'
+        exclude:
+        - variant: standard
+          configuration: Debug
+    runs-on: ${{ matrix.runner }}
+    steps:
+    - name: Install Visual Studio 2017
```
Comment (stinos) [inline .github/workflows/ports_windows.yml:43]: This is rather slow, could this be cached?

File: .github/workflows/ports_windows.yml (hunk 3 above)
Comment (dlech) [reply .github/workflows/ports_windows.yml:55]: It matches the version installed by default  in the `windows-2022` image.

File: .github/workflows/ports_windows.yml (hunk 4 above)
Comment (dlech) [reply .github/workflows/ports_windows.yml:43]: I'm guessing that the installers poke windows registry, so probably not. Maybe there is a way to create a docker container with this pre-installed or something like that? 

(It's already still faster than the current appveyor build though)

**Comment by dlech** [comment]
> but I also think we need at least a POC that remote login is possible.

I was able to successfully log in via RDP with https://github.com/marketplace/actions/debugging-using-reverse-rdp-with-ngrok

Since it requires a personal access token and only allows one session at a time, I don't think it makes sense to commit anything here, but to use it, the basic procedure is to delete all of the other workflow files and all of the jobs except for one in the windows workflow file (so there is only one matrix combination left) and add this action as a step wherever it makes sense (may want it before or after other steps depending on where the failure is).

Also, don't do what I did and set the step to foreground: true and then shut down the remote windows vm without deleting `Delete-This-File-To-Continue.txt`. It caused the CI to hang forever and I couldn't start another run.

**Comment by dpgeorge** [comment]
I'd like to move forward with this PR and get it merged.

@stinos @jimmo do you have any objections given that it is possible to RDP (see above comment)?

**Comment by dpgeorge** [comment]
Will need to merge #10267 before this one.

File: ports/windows/Makefile [hunk 5]
```diff
@@ -92,6 +92,12 @@ include $(TOP)/py/mkrules.mk
 
 RUN_TESTS_SKIP += -e math_fun -e float2int_double -e float_parse -e math_domain_special
 
+ifeq ($(MSYSTEM),MINGW32)
+# compliler optimizations break float formatting
+# https://github.com/micropython/micropython/pull/10267
+RUN_TESTS_SKIP += -e format_ints_doubleprec -e format_modulo2_intbig
```
Comment (stinos) [inline ports/windows/Makefile:98]: Is there a reason to do the skipping here instead of directly in run_tests.py?

**Comment by stinos** [comment]
I gguess this should be fine. Might also make it easier to publish the executables?

File: ports/windows/Makefile (hunk 5 above)
Comment (dpgeorge) [reply ports/windows/Makefile:98]: Merging #10267 will make this change here go away.

**Comment by jimmo** [comment]
@dpgeorge Yep keen to merge! Thanks @dlech

**Comment by dlech** [comment]
It looks like the 2022 builds are failing because of an extra space in the path in the "Get micropython.exe path" step.

```
micropython= D:\a\micropython\micropython\ports\windows\build-dev\Releasex64\micropython.exe
```

There should not be a space between "=" and "D". Since the other VS version builds aren't failing, I'm guessing something must have changed in VS 2022 in the last year since this was working before. I'm pretty sure that is why I put the `.Trim()` call in there, so I'm not sure why we would end up with a space there.

**Comment by dlech** [comment]
I suppose it could have something to do with a different PowerShell version on the windows-2022 VM too.

**Comment by dpgeorge** [comment]
Yes, I'm just trying to fix this now, with the space/Trim.

@dlech sorry but I forced pushed to your branch!  I also rebased on latest master.

**Comment by dpgeorge** [comment]
OK, I think this is working now.  Needed a small tweak to remove the space between the variable name and the variable value.

This new github actions workflow has a lot of jobs... that's mainly due to a large job matrix with different versions of Visual Studio, x86 and x64, debug and release builds.

The VS 2017 jobs are very slow, they seem to get stuck on `choco install visualstudio2017-workload-vctools` for a while, and the whole "install visual studio" step takes about 6 minutes for VS 2017 (2019 and 2022 don't need this extra install step).  Do we need to test against VS 2017?  I'm guessing we do because that's what the current AppVeyor uses.

**Comment by dpgeorge** [comment]
Maybe we can just exclude VS 2017 builds of the standard variant?  That will cut off 2 jobs.  Then there's still 4 jobs for VS 2017, all building the dev variant: x86 debug, x86 release, x64 debug, x64 release.

**Comment by stinos** [comment]
> OK, I think this is working now. Needed a small tweak to remove the space between the variable name and the variable value.

I would think this is the issue described [here](https://github.com/stinos/micropython-wrap/commit/5ded8b4a73a7f19ebf8fd52a836140c973f54641#diff-e1ab4c3cec24f841948eb867754d4ad0b9b6d11eaa6b5bfbe0357aceb0a49effR85) yet you fixed this with only one Trim so not sure.

> This new github actions workflow has a lot of jobs... that's mainly due to a large job matrix with different versions of Visual Studio, x86 and x64, debug and release builds.
> 
> The VS 2017 jobs are very slow, they seem to get stuck on `choco install visualstudio2017-workload-vctools` for a while, and the whole "install visual studio" step takes about 6 minutes for VS 2017 (2019 and 2022 don't need this extra install step). Do we need to test against VS 2017? I'm guessing we do because that's what the current AppVeyor uses.

I always opted to keep this in for backwards compatibility, but by now we're like 7 years further so imo VS2017 is not really crucial anymore. For the variants we could for example build the dev one for Debug and Release, and the standard one only for Release? Likewise only build the VS2019 in Release.

**Comment by dpgeorge** [comment]
> For the variants we could for example build the dev one for Debug and Release, and the standard one only for Release? Likewise only build the VS2019 in Release.

I've now excluded VS 2019 in release mode (the standard variant was already only built in release mode).

The set of jobs are now:

jobs

**Comment by stinos** [comment]
@dlech do you happen to remember the full story for needing https://github.com/micropython/micropython/blob/master/tests/run-tests.py#L586 and https://github.com/micropython/micropython/blob/master/.github/workflows/ports_windows.yml#L141 ? 

Sorry I didn't see this when looking at the PR but that "import/import_file.py" test should already have been excluded by https://github.com/micropython/micropython/blob/master/tests/run-tests.py#L701 but somehow the `if os.name == "nt":` is not `True` in the mingw builds on Github Actions. Don't really understand how (thing runs on Windows after all, and the whole point of having pacman install python3 is that it can be used as-is for running the test suite without needing an 'external' Python installation), but it's pretty annoying because now for each test we want to skip on Windows we have to add it in 2 locations in run-tests. Well, or figure out how to make the CI build run like a local one on Windows; if this doesn't ring any bells for you I'll look into it.

**Comment by stinos** [comment]
> So this is why we need to use the different version of CPython - to get one with Windows line endings so that the generated .exp files match between 

Hmm. I really thought of run-tests has `replace(b"\r\n", b"\n")` in strategic places (both to get micropython and expected test output) just to fix any possible issue there.

Anyway the key issue here is that following the windows README for an MSYS2 installation has none of these problems, and its python version does have `'nt'` for `os.name`.

I'll see if I can get RDP running..

**Pull request merged**